java/client/src/org/openqa/selenium/virtualauthenticator/Credential.java

```diff
@@ -0,0 +1,116 @@
+// Licensed to the Software Freedom Conservancy (SFC) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The SFC licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package org.openqa.selenium.virtualauthenticator;
+
+import java.security.spec.PKCS8EncodedKeySpec;
+import java.util.Base64;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * A credential stored in a virtual authenticator.
+ * @see https://w3c.github.io/webauthn/#credential-parameters
+ */
+public class Credential {
+
+  private final byte[] id;
+  private final boolean isResidentCredential;
+  private final String rpId;
+  private final PKCS8EncodedKeySpec privateKey;
+  private final byte[] userHandle;
+  private final int signCount;
+
+  /**
+   * Creates a non resident (i.e. stateless) credential.
+   */
+  public static Credential createNonResidentCredential(byte[] id, String rpId,
+      PKCS8EncodedKeySpec privateKey, int signCount) {
+    return new Credential(id, /*isResidentCredential=*/false, Objects.requireNonNull(rpId),
+        privateKey, /*userHandle=*/null, signCount);
+  }
+
+  /**
+   * Creates a resident (i.e. stateful) credential.
+   */
+  public static Credential createResidentCredential(byte[] id, String rpId,
+      PKCS8EncodedKeySpec privateKey, byte[] userHandle, int signCount) {
+    return new Credential(id, /*isResidentCredential=*/true, Objects.requireNonNull(rpId),
+        privateKey, Objects.requireNonNull(userHandle), signCount);
+  }
+
+  /**
+   * Creates a credential from a map.
+   */
+  public static Credential fromMap(Map<String, Object> map) {
```

Comment:
You may or may not know this, but Selenium's JSON library will use this method to deserialise a JSON stream to a credential. It'll use `toJson` to go the other way. Neither of those methods need to be `public`: if you don't want them to be visible outside this package, they don't need to be.

Which is a really long way of saying that this method could be package protected and everything should still work. Do you want this to be part of the public API?

Reply:
I think we still need this to be public to [call it from `RemoteWebDriver.java`](https://github.com/SeleniumHQ/selenium/pull/7842/files#diff-fa76272f866d8067bb11a32f27773026R1088) since they're in different packages.

> You may or may not know this, but Selenium's JSON library will use this method to deserialise a JSON stream to a credential.

I had no idea -- are you saying we can somehow wire `Response.getValue()` to do this for us? That would probably be more elegant but don't know how to do it.

Reply:
It doesn't do it yet. I really should make that happen. In the Grid code, we have a helper `Values` class that _does_ do this.